### PR TITLE
Bootstrapping for homomorphic evaluator

### PR DESCRIPTION
**Pull request by crockeea**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by crockeea** [comment]
CI will fail until the [latticpp PR](https://github.com/awslabs/aws-cppwrapper-lattigo/pull/6) is merged.

File: src/hit/api/evaluator/homomorphic.cpp [hunk 1]
```diff
@@ -308,4 +342,16 @@ namespace hit {
     void HomomorphicEval::relinearize_inplace_internal(CKKSCiphertext &ct) {
         relinearize(get_evaluator(), ct.backend_ct, relin_keys, ct.backend_ct);
     }
+
+    CKKSCiphertext HomomorphicEval::bootstrap_internal(const CKKSCiphertext &ct, bool rescale_for_bootstrapping) {
```
Comment (bryce-shang) [inline src/hit/api/evaluator/homomorphic.cpp:346]: Is it an internal method? I cannot find its usage.

File: src/hit/api/evaluator/homomorphic.cpp [hunk 2]
```diff
@@ -236,6 +256,20 @@ namespace hit {
...
+        }
+
+        if (backend_bootstrapper.get() == nullptr || backend_bootstrapper->params != context->params) {
+            ParameterizedLattigoType<Bootstrapper> *tmp =
+                new ParameterizedLattigoType<Bootstrapper>(
+                    newBootstrapper(context->params, context->btp_params.value(), btp_keys), context->params);
+            backend_bootstrapper.reset(tmp);
+        }
+        return backend_bootstrapper->object;
+    }
```
Comment (bryce-shang) [inline src/hit/api/evaluator/homomorphic.cpp:271]: Need lock?

File: src/hit/api/evaluator/homomorphic.cpp (hunk 1 above)
Comment (crockeea) [reply src/hit/api/evaluator/homomorphic.cpp:346]: `bootstrap_internal` is the evaluator-specific bootstrapping function. The user API is here: https://github.com/awslabs/homomorphic-implementors-toolkit/blob/wip-bootstrapping-homomorphic/src/hit/api/evaluator.cpp#L382

File: src/hit/api/evaluator/homomorphic.cpp (hunk 2 above)
Comment (crockeea) [reply src/hit/api/evaluator/homomorphic.cpp:271]: Not needed here. `backend_bootstrapper` is a `thread_specific_ptr`, which means I don't have to worry about two threads using the same `backend_bootstrapper`. This code just ensures that a `backend_bootstrapper` exists for each thread.

**Review by bryce-shang — APPROVED** [review]

**Pull request merged**